### ISSUE TITLE
Chapter-7 Request

### ISSUE THREAD

**Issue by victorvdo** [ageron/handson-ml#217]
I love the book!!!

One request, not issue, for Chapter 7: It'd be great if there is an example of building a ensemble of different kinds of "regressions" such SVM regression, random forest regression, etc.... using the housing data.

The book has only the classifier ensemble example. Thank you.

**Comment by ageron** [comment]
Hi Victor,

Thanks for your kind words! :)

For regression, one simple option is to create multiple regressors and do a weighted sum of their predictions (e.g., weighted by their r2 score on the validation set). Another approach is to use stacking (checkout the solution to exercise 7.9: it shouldn't be hard to switch it from classification to regression).

Here's an implementation of the first option:

```python
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_diabetes
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor, ExtraTreesRegressor
from sklearn.base import BaseEstimator, RegressorMixin, clone

X, y = load_diabetes(return_X_y=True)

X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=42)

class WeightedMeanEnsemble(BaseEstimator, RegressorMixin):
    def __init__(self, estimators, weights=None):
        self.estimators = estimators
        self.weights = weights
​
    def fit(self, X, y, **params):
        self.estimators_ = [clone(reg).fit(X, y, **params)
                            for name, reg in self.estimators]
        return self
​
    def predict(self, X):
        predictions = np.array([estimator.predict(X) for estimator in self.estimators_])
        return np.average(predictions, axis=0, weights=self.weights)

weighted_mean_ensemble = WeightedMeanEnsemble([
    ("ridge", Ridge(random_state=42)),
    ("rnd_forest_reg", RandomForestRegressor(random_state=42)),
    ("extra_trees_reg", ExtraTreesRegressor(random_state=42)),
])

weighted_mean_ensemble.fit(X_train, y_train)

print(weighted_mean_ensemble.score(X_val, y_val))
# 0.47083442135168646

weights = [reg.score(X_val, y_val) for reg in weighted_mean_ensemble.estimators_]
print(weights)
# [0.4384009113704955, 0.4084100811626866, 0.4379750225843305]

print(weighted_mean_ensemble.score(X_test, y_test))
# 0.47083442135168646

weighted_mean_ensemble.weights = weights

print(weighted_mean_ensemble.score(X_test, y_test))
# 0.47130267314071506  = very slight improvement

[reg.score(X_test, y_test) for reg in weighted_mean_ensemble.estimators_]
# [0.4384009113704955, 0.4084100811626866, 0.4379750225843305]
```

Notice that the ensemble performs better than all the individual regressors, and the weighted ensemble performs better than the first ensemble (which gives all estimators the same weight).

Hope this helps!
